Add "name" variable support for tf.einsum

yongtang opened the pull request:
This fix adds "name" variable support for tf.einsum

This fix fixes #11501

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add "name" variable support for tf.einsum

This fix adds "name" variable support for tf.einsum

This fix fixes #11501

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

vrv commented on tensorflow/python/ops/special_math_ops.py:
```diff
@@ -82,7 +82,7 @@ def lbeta(x, name='lbeta'):
 
 
-def einsum(equation, *inputs):
+def einsum(equation, *inputs, **kwargs):
```
Is it possible to avoid the use of the **kwargs and just add name=None instead?

josh11b replied:
It appears to be unavoidable, see https://stackoverflow.com/questions/20961100/passing-default-parameters-with-varargs-in-function-in-python

yongtang commented:
@vrv Thanks for the review. If we use `name=None` at the end then it will thrown out a syntax error:
```
  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/ad1e09741bb4109fbc70ef8216b59ee2/execroot/tensorflow/bazel-out/local-opt/bin/tensorflow/python/special_math_ops_test.runfiles/org_tensorflow/tensorflow/python/ops/special_math_ops.py", line 85
    def einsum(equation, *inputs, name=None):
                                     ^
SyntaxError: invalid syntax
```

So I used `**kwargs` to get around it.

vrv commented on tensorflow/python/ops/special_math_ops.py:
```diff
@@ -151,70 +152,76 @@ def einsum(equation, *inputs):
+  name = kwargs.pop("name", None)
+  if kwargs:
+    key = list(kwargs.keys())[0]
+    raise TypeError("'{}' is an invalid keyword argument "
```
Should we raise an error for all the keys passed, not just the first one?

vrv commented:
@tensorflow-jenkins test this please

yongtang committed:
Update API Golden for api_compatibility_test

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @vrv for the review. The PR has been updated so that all invalid argument keys would be printed in case of error. A test case has been added as well.

I also updated with
```
   $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```
so that api_compatibility_test could pass on Jenkins.

Please take a look and let me know if there are any issues.

Commit:
reflow

Commit:
Match indentation

vrv commented:
Thanks!  This looks good.

I'm not sure if this qualifies as something that needs API review -- presumably it does and it should be fine.

@tensorflow-jenkins test this please

yongtang committed:
Change the test case to better cover the scenario.

This commit changes the test case from
```
invalid1="invalid1", invalid2="invalid2"
```
to
```
invalid1="value1", invalid2="value2"
```
to make sure that the error message will show the keyward arguments,
not the value of the argument.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

josh11b commented:
Fine for API review.

rmlarsen commented:
@tensorflow-jenkins test this please

rmlarsen commented:
@yongtang It looks like you need to fix the expected error message in your test:

Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/tensorflow-pull-requests-cpu-python3/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/workspace/bazel-out/local-py3-opt/bin/bazel_pip/tensorflow/python/special_math_ops_test.runfiles/org_tensorflow/bazel_pip/tensorflow/python/ops/special_math_ops_test.py", line 251, in test_invalid_keyword_arguments
    invalid1="value1", invalid2="value2")
**AssertionError: "invalid keyword arguments for this function: invalid1, invalid2" does not match "invalid keyword arguments for this function: invalid2, invalid1"**

yongtang committed:
Fix the expected error message in the tests.

This fix sorts the list so that error message is repeatable.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@rmlarsen Thanks for the review. The Jenkins error for was because dict in python is not ordered when iterate through. I have sorted the error message so that the output is repeatable. The PR has been updated now.

rmlarsen commented:
@yongtang thanks for the quick fix. Yeah, iterating over and unordered container is asking for trouble :-)
@tensorflow-jenkins test this please
